### DISCUSSION TITLE
In HardwareTimer library, generic update callback should clear update interrupt flag (UIF) automatically

### DISCUSSION THREAD

**Discussion by AndrewBCN** [stm32duino/Arduino_Core_STM32#1401] [Q&A] [answered]
Hello,
This is a suggestion for a small improvement in the HardwareTimer library, which I have been using.
In the HardwareTimer library, in file https://github.com/stm32duino/Arduino_Core_STM32/blob/master/cores/arduino/HardwareTimer.cpp
specifically in the following function:
`
/**  * @brief  Generic Update (rollover) callback which will call user callback  * @param  htim: HAL timer handle  * @retval None  */void HardwareTimer::updateCallback(TIM_HandleTypeDef *htim){  if (!htim) {    Error_Handler();  }
  timerObj_t *obj = get_timer_obj(htim);  HardwareTimer *HT = (HardwareTimer *)(obj-&gt;__this);
  if (HT-&gt;callbacks[0]) {    HT-&gt;callbacks[0]();  }}`

One can see that the Update Interrupt Flag (UIF) for the timer in question is not cleared by the generic update callback function.

My suggestion is that the UIF flag should be cleared and that this should be documented in wiki, that there is no need for the user callback function to manually clear this flag.

The purpose of having the flag cleared automatically by the callback function is that the flag can then be tested by user code to check if the user callback function has executed or not, consequently we know if there was an update event and that it has already been handled by the user callback function.

I have checked the programming manual and the UIF is not cleared by hardware, only by software.

**Comment by ABOSTM** [comment, marked as answer]
> One can see that the Update Interrupt Flag (UIF) for the timer in question is not cleared by the generic update callback function.

In fact, UIF flag is cleared:
HardwareTimer library is based on STM32 HAL driver, and this is the place where UIF flag is cleared.
Below is an example of call stack when Timer interrupt occurs:
```
  TIM1_IRQHandler()
   |-> HAL_TIM_IRQHandler()
        |->HAL_TIM_PeriodElapsedCallback()
              |->HardwareTimer::updateCallback(htim)
```
Flag is cleared in HAL_TIM_IRQHandler() before calling HAL_TIM_PeriodElapsedCallback()
https://github.com/stm32duino/Arduino_Core_STM32/blob/75d653197e3362b015def7a6d1a0a6898d2d0ace/system/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.c#L3860

**Comment by AndrewBCN** [comment]
Ah, my bad!
Thank you very much, excellent explanation providing an insight in how the Hardware Timer library works. I am going to add a note in the wiki in the Hardware Timer section, mentioning that the UIF flag is automatically cleared by the HAL interrupt handler, before the user callback function is executed.
